spring-data-jdbc/src/main/java/org/springframework/data/jdbc/repository/aot/JdbcCodeBlocks.java

```diff
@@ -0,0 +1,789 @@
+						builder.addStatement("$T $L = getRowMapperFactory().getResultSetExtractor($S)", ResultSetExtractor.class,
+								resultSetExtractor, resultSetExtractorRef);
+					}
+				}
+
+				if (StringUtils.hasText(resultSetExtractor)) {
+
+					builder.addStatement("return ($T) getJdbcOperations().query($L, $L, $L)", queryResultType, queryVariableName,
+							parameterSourceVariableName, resultSetExtractor);
+
+					return builder.build();
+				}
+
+				boolean dynamicProjection = StringUtils.hasText(context.getDynamicProjectionParameterName());
+				Object queryResultTypeRef = dynamicProjection ? context.getDynamicProjectionParameterName() : queryResultType;
+
+				if (queryMethod.isCollectionQuery() || queryMethod.isSliceQuery() || queryMethod.isPageQuery()) {
+
+					builder.addStatement("$1T $2L = ($1T) getJdbcOperations().query($3L, $4L, new $5T<>($6L))", List.class,
+							result, queryVariableName, parameterSourceVariableName, RowMapperResultSetExtractor.class, rowMapper);
+
+					if (queryMethod.isSliceQuery() || queryMethod.isPageQuery()) {
+
+						String pageable = context.getPageableParameterName();
+
+						builder.addStatement(
+								"$1T $2L = ($1T) convertMany($3L, %s)".formatted(dynamicProjection ? "$4L" : "$4T.class"), List.class,
+								context.localVariable("converted"), result, queryResultTypeRef);
+
+						if (queryMethod.isPageQuery()) {
+
+							builder.addStatement("return $1T.getPage($2L, $3L, $4L)", PageableExecutionUtils.class,
+									context.localVariable("converted"), pageable, context.localVariable("countAll"));
+						} else {
+
+							builder.addStatement("boolean $1L = $2L.isPaged() && $3L.size() > $2L.getPageSize()",
+									context.localVariable("hasNext"), pageable, context.localVariable("converted"));
+
+							builder.addStatement("return new $1T($2L ? $3L.subList(0, $4L.getPageSize()) : $3L, $4L, $2L)",
+									SliceImpl.class, context.localVariable("hasNext"), context.localVariable("converted"), pageable);
+						}
+
+						return builder.build();
+					}
+
+					builder.addStatement("return ($T) convertMany($L, %s)".formatted(dynamicProjection ? "$L" : "$T.class"),
+							context.getReturnTypeName(), result, queryResultTypeRef);
+				} else if (queryMethod.isStreamQuery()) {
+
+					builder.addStatement("$1T $2L = getJdbcOperations().queryForStream($3L, $4L, $5L)", Stream.class, result,
+							queryVariableName, parameterSourceVariableName, rowMapper);
+					builder.addStatement("return ($T) convertMany($L, $T.class)", context.getReturnTypeName(), result,
+							queryResultTypeRef);
+				} else {
+
+					builder.addStatement("$T $L = queryForObject($L, $L, $L)", Object.class, result, queryVariableName,
+							parameterSourceVariableName, rowMapper);
+
+					if (Optional.class.isAssignableFrom(context.getReturnType().toClass())) {
+						builder.addStatement(
+								"return ($1T) $1T.ofNullable(convertOne($2L, %s))".formatted(dynamicProjection ? "$3L" : "$3T.class"),
```

Comment:
do we need that `($1T)` cast to do satisfy the compiler? `(Optional) Optional.ofNullable(...)` looks strange though.